### PR TITLE
nixos/cdi.dynamic.nvidia: expose driverLink

### PR DESCRIPTION
**Pull request by SomeoneSerge**
## Description of changes

A quick follow-up to https://github.com/NixOS/nixpkgs/pull/284507. I'd hate to merge something with that many FIXME comments, but I'm not going to have the time to implement these in a while yet

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: nixos/modules/services/hardware/nvidia-container-toolkit-cdi-generator/cdi-generate.nix [hunk 1]
```diff
@@ -1,37 +1,58 @@
...
+  runtimeShell,
+  writeScriptBin,
+}:
+let
   mountOptions = { options = ["ro" "nosuid" "nodev" "bind"]; };
   mounts = [
-    { hostPath = "${lib.getBin config.hardware.nvidia.package}/bin/nvidia-cuda-mps-control";
+    # FIXME: Making /usr mounts optional
+    { hostPath = lib.getExe' nvidia-driver "nvidia-cuda-mps-control";
       containerPath = "/usr/bin/nvidia-cuda-mps-control"; }
```
Comment (SomeoneSerge) [inline nixos/modules/services/hardware/nvidia-container-toolkit-cdi-generator/cdi-generate.nix:16]: I only now realized that this isn't going to be preserved by any of the formatters we've got, but I wanted to keep the diff small(er)

**Comment by ereslibre** [comment]
> I'd hate to merge something with that many FIXME comments, but I'm not going to have the time to implement these in a while yet

Let me have a look later today. I can try to help with those FIXME comments on a separate PR. This has been my first NixOS module contribution and I am not fully aware of all the details, but I can try to move that forward.

**Comment by ereslibre** [comment]
I just tried this PR, I got the following error:

```
$ sudo nixos-rebuild --flake '.#hulk' switch
warning: Git tree '/home/ereslibre/projects/homelab' is dirty
building the system configuration...
warning: Git tree '/home/ereslibre/projects/homelab' is dirty
activating the configuration...
setting up /etc...
reloading user units for ereslibre...
restarting sysinit-reactivation.target
warning: the following units failed: nvidia-control-devices.service

× nvidia-control-devices.service
     Loaded: loaded (/etc/systemd/system/nvidia-control-devices.service; enabled; preset: enabled)
     Active: failed (Result: exit-code) since Tue 2024-02-27 21:59:17 CET; 227ms ago
   Duration: 10ms
    Process: 118225 ExecStart=/nix/store/jvd8kqg786b06vp8rsmv8fhaa24d6y0a-nvidia-x11-550.54.14-6.6.18-bin/bin/nvidia-smi (code=exited, status=18)
   Main PID: 118225 (code=exited, status=18)
         IP: 0B in, 0B out
        CPU: 10ms

Feb 27 21:59:17 hulk systemd[1]: Started nvidia-control-devices.service.
Feb 27 21:59:17 hulk nvidia-smi[118225]: Failed to initialize NVML: Driver/library version mismatch
Feb 27 21:59:17 hulk nvidia-smi[118225]: NVML library version: 550.54
Feb 27 21:59:17 hulk systemd[1]: nvidia-control-devices.service: Main process exited, code=exited, status=18/n/a
Feb 27 21:59:17 hulk systemd[1]: nvidia-control-devices.service: Failed with result 'exit-code'.
warning: error(s) occurred while switching to the new configuration
```

**Review by ereslibre — APPROVED** [review]
Okay, after a reboot it works fine. I guess the driver did update and I had to reboot to activate it. LGTM.

**Pull request merged**